### ISSUE TITLE
Regent/Legion Pruning Use of Empty Regions

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#434]
Sometimes empty logical regions get made. It happens (especially with static control replication). We should do better at avoiding the costs of them as they add to runtime overhead. 

Two proposed changes:
1. The runtime re-writes region requirements for empty logical regions to NO_ACCESS
2. Have the compiler recognize when regions have been rewritten to NO_ACCESS to avoid using them.

**Comment by streichler** [comment]
Will switching the region requirements to NO_ACCESS cause problems in the task body when it makes an accessor with read/write/reduce privileges?

**Comment by elliottslaughter** [comment]
The C API accessors don't use privileges:

https://github.com/StanfordLegion/legion/blob/master/runtime/legion/legion_c.h#L3266-L3269

I think we have a guard so that we should be ok accepting null pointers if the volume of the region is 0. At the moment the compiler will still request an accessor, and I don't think that should change because various things in the compiler assume that the variables are set up unconditionally.

**Comment by elliottslaughter** [comment]
One thing to be careful about if you do this is that you will need to modify the runtime so that a task can call another task with READ_WRITE privilege if the region is empty, even if the parent has NO_ACCESS. Otherwise you will break any transitive task calls in Regent.

**Comment by lightsighter** [comment]
@elliottslaughter The only regions that can inherit privileges from an empty sub-region must also be empty and therefore will also be re-written to NO_ACCESS.

**Comment by elliottslaughter** [comment]
That's fine, as long as you do your rewrite to NO_ACCESS before checking privileges.

**Comment by lightsighter** [comment]
Yes, that is when it would have to be done.

**Comment by elliottslaughter** [comment]
I think option (2) isn't really feasible under DCR since all launches are (or should be) index launches. Is this still an issue and is (1) a viable approach?

**Comment by lightsighter** [comment]
Yes, this is still potentially an issue under DCR. It's the emptiness of the index spaces for the logical regions themselves, not the index space launches that is the issue.
